Source/DafnyCore/AST/Expressions/ExtendedPattern.cs

```diff
@@ -15,4 +22,142 @@ public abstract class ExtendedPattern : INode {
 
   public IEnumerable<INode> DescendantsAndSelf =>
     new[] { this }.Concat(Children.OfType<ExtendedPattern>().SelectMany(c => c.DescendantsAndSelf));
+
+  public abstract void Resolve(Resolver resolver, ResolutionContext resolutionContext,
+    Type sourceType, bool isGhost, bool mutable,
+    bool inPattern, bool inDisjunctivePattern);
+
+  public abstract IEnumerable<(BoundVar var, Expression usage)> ReplaceTypesWithBoundVariables(Resolver resolver,
+    ResolutionContext resolutionContext);
+
+  /*
+  *  Ensures that all ExtendedPattern held in NestedMatchCase are linear
+  *  Uses provided type to determine if IdPatterns are datatypes (of the provided type) or variables
+  *  pat could be
+  *  0 - A DisjunctivePattern
+  *  1 - An IdPattern (without argument) at base type
+  *  2 - A LitPattern at base type
+  *  3* - An IdPattern at tuple type representing a tuple
+  *  3 - An IdPattern at datatype type representing a constructor of type
+  *  4 - An IdPattern at datatype type with no arguments representing a bound variable
+  */
+  public void CheckLinearExtendedPattern(Type type, ResolutionContext resolutionContext, Resolver resolver) {
+    if (type == null) {
```

Comment:
That doesn't seem possible to me. Note this is an existing check, this method was moved from `Resolver.cs`. In a follow-up PR, I will merge `CheckLinearExtendedPattern` and `ExtendedPattern.Resolve` and maybe this can go then.